{"cells":[{"cell_type":"code","source":"# RA, 2020-09-29\n# How are paired reads oriented in the reference genome?","metadata":{"tags":[],"cell_id":"00000-55951b2a-7a13-4dee-b4ee-82640b5874e5","output_cleared":false},"outputs":[],"execution_count":8},{"cell_type":"code","metadata":{"tags":[],"cell_id":"00000-94aac15a-43de-4228-8e1b-921e56cee572","output_cleared":false},"source":"from pathlib import Path\n\ntry:\n    __file__\nexcept:\n    import os\n    __file__ = Path(os.getcwd()) / \"unknown\"\n\n# Reference genome file\nfa = Path(__file__).parent.parent / \"input/data_small/genome.chr22.5K.fa\"","execution_count":9,"outputs":[]},{"cell_type":"code","source":"# String operators\nforward = (lambda s: s)\nreverse = (lambda s: ''.join({'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}[c] for c in s[-1::-1]))","metadata":{"tags":[],"cell_id":"00002-0e86ded5-d04a-45c8-b09b-23f084aba4db","output_cleared":false},"outputs":[],"execution_count":10},{"cell_type":"code","source":"def find(pattern, template):\n    \"\"\"\n    Find all occurrences of pattern string in template string.\n    Note: returns a list of starting positions with 1- based indexing.\n    \"\"\"\n    import re\n    return [(m.start() + 1) for m in re.finditer(\"(?=\" + pattern + \")\", template)]","metadata":{"tags":[],"cell_id":"00003-030a7e43-5c6a-42d7-88b6-4af0630665dd","output_cleared":false},"outputs":[],"execution_count":12},{"cell_type":"code","source":"from Bio import SeqIO\ntemplate = str(SeqIO.read(fa, format='fasta').seq)","metadata":{"tags":[],"cell_id":"00001-ceb92dd2-ec0c-4049-8989-dc688454fadf","output_cleared":false},"outputs":[],"execution_count":11},{"cell_type":"code","source":"# This is from output_tiny_30xCov1.fq: 22_5K-1168/1\nr1 = \"TCTGGGCCTCCCAACCCTGAGTTTTTATAATAGGCCCCAGGCCAGGTGGTTAACAGAGGTCTGGGGCATTGCAGGGGGACAGAGGAGGACATATGTCCCTATTGGCCATTGTAGAGTCCCTTCCA\"\n# This is from output_tiny_30xCov2.fq: 22_5K-1168/2\n# It is also the same as 22_5K-880/1 and 22_5K-712/2\nr2 = \"CTGGGTGACAGAAGGAGACTCTGTCTAGGAAAAACAAAAATTTTGCAGAAAATATGTTGAGACAGGATCTATGTTGCCCAGGCTGGTCTTGAACTCCTGGGCTCAAGCAATCCTTCTGCTTCCGA\"","metadata":{"tags":[],"cell_id":"00004-d44aa140-17c5-4619-9e61-33d225606634","output_cleared":false},"outputs":[],"execution_count":13},{"cell_type":"code","source":"# Expect that one of the paired reads will be matched as reverse-complement\n# Note: if the reverse-complement R of a read r is matched, it shows up in the SAM file as R (and not r)\nfor (op, op_name) in [(forward, \"forward\"), (reverse, \"reverse-complement\")]:\n    print(F\"Matches for {op_name} string\")\n    print(\"    Read 1:\", find(op(r1), template))\n    print(\"    Read 2:\", find(op(r2), template))\n\n    # print(op(r1))\n    # print(op(r2))","metadata":{"tags":[],"cell_id":"00006-535c3aa3-99d4-4091-825d-5cad26d07b10","output_cleared":false},"outputs":[{"name":"stdout","text":"Matches for forward string\n    Read 1: []\n    Read 2: [2364]\nMatches for reverse-complement string\n    Read 1: [2674]\n    Read 2: []\n","output_type":"stream"}],"execution_count":14}],"nbformat":4,"nbformat_minor":2,"metadata":{"orig_nbformat":2,"deepnote_notebook_id":"c27896af-8772-4a7e-9a84-181f3be85583","deepnote_execution_queue":[]}}