{"cells":[{"cell_type":"code","metadata":{"cell_id":"00000-9bfa2b3b-a336-45e3-9d1e-42c74370475a"},"source":"# Comp Biomed \n# Notes for call 20200927\n# RA, ...","execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Note: maybe the repo \nhttps://gitlab.inf.ethz.ch/COURSE-BIOMED20/team03\n can be pulled to this environment, but I haven't done so yet (RA)","metadata":{"tags":[],"cell_id":"00001-a46b03cb-e3eb-4f21-9729-05a39f1e6cae"}},{"cell_type":"markdown","source":"# Project 1 description\n[PDF](./project1_description.pdf)\n\n# Overview\n - decide on and implement a genome indexing strategy\n - decide on and implement a read alignment method\n - define and implement routines to read input data and generate correctly formated output\n - define and apply evaluation metrics to judge the quality of your tool\n\n## 1: read\nDevise and implement a reader for the input data.\n - fa: [wiki](https://en.wikipedia.org/wiki/FASTA_format)\n - fq: [wiki](https://en.wikipedia.org/wiki/FASTQ_format)\n - sam: [wiki](https://en.wikipedia.org/wiki/SAM_(file_format)), [samtools](https://github.com/samtools/hts-specs/blob/master/SAMv1.pdf)\n \nscalability by amount of samples, length of whole ref, length of p\n\nscalability in space used and time consumed\n**(& performance: Jannik)**\n\n## 2: write\nWe expect you to generate your output alignments in standard [SAM format](https://github.com/samtools/hts-specs/blob/master/SAMv1.pdf).\nThe goal of this work package is to understand the output format specification and design and implement\ncode that generates correctly formatted output for each of the alignments.\n\n## 3: index (*)\nDevise and implement a method that generates an index on the input\ndata. \nIdeally, this index can be stored on disk and re-used at a later time.\n\n\nBWT+FM\n**(Luca)**\n\n## 4: alignment (*)\nDevelop an alignment software tool that can find the position(s) of one\nor many reads provided in fastq format in a fully indexed genome.\n**(Hendrik)**\n\n## Quality of alignment\n**(Roman)**\n\n## Submission\nCode needs to be executable in the provided **conda** environment. \nIf additional packages have been approvedfor use by the TA team, please provide a conda recipe. \nAny solution that is not using Python needs to beprovided as a Docker container and \na detailed README on how to execute the code.\nPlease understandthat for practical purposes any code that a qualified TA \nis unable to get running within 10-15 minutesdue to missing information, can not be graded.\n\n# Github\n[GitHub](https://github.com/numpde/Team03)\n\n# Next call\nSun, Oct 11, 11am","metadata":{"tags":[],"cell_id":"00002-06b67402-6633-4ec9-98d3-3b9e00cf3b36"}},{"cell_type":"markdown","source":"","metadata":{"tags":[],"cell_id":"00003-d16a24d3-44a6-4be4-ba33-66075ae273f8"}}],"nbformat":4,"nbformat_minor":2,"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.6.3"},"deepnote_notebook_id":"8724d716-346a-482c-8284-b39346d07365","deepnote_execution_queue":[]}}