{"cells":[{"cell_type":"code","metadata":{"cell_id":"00000-42ffb411-07ad-4d7b-947d-edb499cd5b4f"},"source":"# Comp Biomed \n# Notes for call 20200927\n# RA, ...","execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Note: maybe the repo \nhttps://gitlab.inf.ethz.ch/COURSE-BIOMED20/team03\n can be pulled to this environment, but I haven't done so yet (RA)","metadata":{"tags":[],"cell_id":"00001-fc7bc517-13f0-45c5-ab49-f0e7a1e68266"}},{"cell_type":"markdown","source":"# Project 1 description\n[PDF](./project1_description.pdf)\n\n# Overview\n - decide on and implement a genome indexing strategy\n - decide on and implement a read alignment method\n - define and implement routines to read input data and generate correctly formated output\n - define and apply evaluation metrics to judge the quality of your tool\n\n# 1: read\nDevise and implement a reader for the input data.\n**(& performance: Jannik)**\n\n# 2: write\nWe expect you to generate your output alignments in standard [SAM format](https://github.com/samtools/hts-specs/blob/master/SAMv1.pdf).\nThe goal of this work package is to understand the output format specification and design and implement\ncode that generates correctly formatted output for each of the alignments.\n\n# 3: index (*)\nDevise and implement a method that generates an index on the input\ndata. \nIdeally, this index can be stored on disk and re-used at a later time.\n**(Luca)**\n\n# 4: alignment (*)\nDevelop an alignment software tool that can find the position(s) of one\nor many reads provided in fastq format in a fully indexed genome.\n**(Hendrik)**\n\n# Quality of alignment\n**(Roman)**\n\n# Submission\nCode needs to be executable in the provided **conda** environment. \nIf additional packages have been approvedfor use by the TA team, please provide a conda recipe. \nAny solution that is not using Python needs to beprovided as a Docker container and \na detailed README on how to execute the code.\nPlease understandthat for practical purposes any code that a qualified TA \nis unable to get running within 10-15 minutesdue to missing information, can not be graded.","metadata":{"tags":[],"cell_id":"00001-01724f41-bd0c-49ae-95c3-3ff0623fc38d"}},{"cell_type":"markdown","source":"## File types\n\n - fa: [wiki](https://en.wikipedia.org/wiki/FASTA_format)\n - fq: [wiki](https://en.wikipedia.org/wiki/FASTQ_format)\n - sam: [wiki](https://en.wikipedia.org/wiki/SAM_(file_format)), [samtools](https://github.com/samtools/hts-specs/blob/master/SAMv1.pdf)\n\n","metadata":{"tags":[],"cell_id":"00004-4c3e4a12-c958-4843-a155-21ba29581c25"}},{"cell_type":"markdown","source":"","metadata":{"tags":[],"cell_id":"00004-8b7b8da1-54c6-4403-83fb-a9193122eb50"}}],"nbformat":4,"nbformat_minor":2,"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.6.3"},"deepnote_notebook_id":"71faaf85-07d2-476a-a015-330d39a34158","deepnote_execution_queue":[]}}